### DISCUSSION TITLE
nextjs 12.2 middleware : ReferenceError: self is not defined

### DISCUSSION THREAD

**Discussion by sylvain-guehria** [vercel/next.js#38223] [Help]
I am using a boilerplate with nextjs 12.2.

I want to use the middleware so I added a middleware.ts in the page folder, as simple as that. But the middleware does not work and the build of the app crashes with the error :
> Build error occurred
ReferenceError: self is not defined
at Object.<anonymous> (/home/sysy73/Documents/projet/future-of-work/talent-frontend/.next/server/pages/middleware.js:1:1)
at Module._compile (node:internal/modules/cjs/loader:1105:14)
at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
at Module.load (node:internal/modules/cjs/loader:981:32)
at Function.Module._load (node:internal/modules/cjs/loader:822:12)
at Module.require (node:internal/modules/cjs/loader:1005:19)
at require (node:internal/modules/cjs/helpers:102:18)
at Object.requirePage (/home/sysy73/Documents/projet/future-of-work/talent-frontend/node_modules/next/dist/server/require.js:61:12)
at /home/sysy73/Documents/projet/future-of-work/talent-frontend/node_modules/next/dist/server/load-components.js:56:50
at processTicksAndRejections (node:internal/process/task_queues:96:5) {
type: 'ReferenceError'
}

I cannot find a way to fix this error in thedoc nor with google.

the boilerplate is this one (I upgraded next to 12.2):
https://github.com/ixartz/Next-js-Boilerplate

Can you tell me more about this self variable ?

**Comment by icyJoseph** [comment]
Hi,

middleware no longer goes in the page folder... it is a sibling to the pages folder.

```
.
├── README.md
├── components
├── hooks
├── jest.config.js
├── middleware.ts
├── next-env.d.ts
├── next.config.js
├── package.json
├── pages
├── public
├── styles
├── tsconfig.json
└── yarn.lock
```

Please refer to, https://nextjs.org/docs/messages/middleware-upgrade-guide#no-nested-middleware:

> Define a single Middleware file next to your pages folder

**Reply by eAmin** [reply]
Hi

I don't have any nested middleware, but I get the same error, Next.js 12.2.4

**Reply by AuthorOfTheSurf** [reply]
@eAmin check your middleware file, it should now be called `middleware.ts` (or `.js`) without an underscore

You may be getting this error if it's still named `_middleware.ts` which corresponds to older versions of next.js. I briefly saw this error myself as I was figuring out how to name and place the file correctly.

> You should declare one single Middleware file in your application, which should be located next to the pages directory and named without an _ prefix. Your Middleware file can still have either a .ts or .js extension.

IMO they should just put a screenshot in the docs due to the major changes in location and filename

**Reply by eAmin** [reply]
Thanks

Yes, I fixed the problem in my project
The problem was that I had placed the middleware file in the **pages** folder, but it should have been **placed next to this folder** and not inside of it.

I was more surprised why there was no problem in the development mode and it only gave an error when building the project and I think it should be fixed or maybe it is already fixed.

**Reply by syd2011** [reply]
same as me. confused
